msgpack-core/src/main/java/org/msgpack/core/MessageUnpacker.java

```diff
@@ -169,195 +169,121 @@ public long getTotalReadBytes()
-        }
-    }
-
     /**
-     * Relocate the cursor position so that it points to the real buffer
-     *
-     * @return true if it succeeds to move the cursor, or false if there is no more buffer to read
-     * @throws IOException when failed to retrieve next buffer
+     * Get the next buffer without changing the position
+     * @return
+     * @throws IOException
      */
-    private boolean ensureBuffer()
+    private MessageBuffer getNextBuffer()
             throws IOException
     {
-        while (buffer != null && position >= buffer.size()) {
-            // Fetch the next buffer
-            int bufferSize = buffer.size();
-            position -= bufferSize;
-            totalReadBytes += bufferSize;
-            buffer = takeNextBuffer();
+        MessageBuffer next = in.next();
+        if (next == null) {
+            throw new MessageInsufficientBufferException();
         }
-        return buffer != null;
+        totalReadBytes += buffer.size();
+        return next;
     }
 
-    private MessageBuffer takeNextBuffer()
+    private void nextBuffer()
             throws IOException
     {
-        if (reachedEOF) {
-            return null;
-        }
-
-        MessageBuffer nextBuffer = null;
-        if (secondaryBuffer == null) {
-            nextBuffer = in.next();
-        }
-        else {
-            nextBuffer = secondaryBuffer;
-            secondaryBuffer = null;
-        }
-
-        if (nextBuffer == null) {
-            reachedEOF = true;
-        }
-        return nextBuffer;
+        buffer = getNextBuffer();
+        position = 0;
     }
 
     /**
-     * Ensure that the buffer has the data of at least the specified size.
-     *
-     * @param byteSizeToRead the data size to be read
-     * @return if the buffer can have the data of the specified size returns true, or if the input source reached an EOF, it returns false.
+     * Returns a short size buffer (upto 8 bytes) to read a number value
+     * @param readLength
+     * @return
      * @throws IOException
+     * @throws MessageInsufficientBufferException If no more buffer can be acquired from the input source for reading the specified data length
      */
-    private boolean ensure(int byteSizeToRead)
+    private MessageBuffer prepareNumberBuffer(int readLength)
             throws IOException
     {
-        if (byteSizeToRead == 0) {
-            return true;
-        }
-
-        if (!ensureBuffer()) {
-            return false;
-        }
-
-        // The buffer contains the data
-        if (position + byteSizeToRead <= buffer.size()) {
-            // OK
-            return true;
+        int remaining = buffer.size() - position;
+        if (remaining >= readLength) {
+            // When the data is contained inside the default buffer
+            nextReadPosition = position;
+            position += readLength;  // here assumes following buffer.getXxx never throws exception
+            return buffer; // Return the default buffer
```

Comment:
[trivial] Why don't we return `buffer.slice(position, readLength)` so that we can remove `nextReadPosition`?